src/types.ts

```diff
@@ -0,0 +1,160 @@
+export type ErrorLike = {
+  message: string;
+  key: string;
+};
+
+export type Callback = (error?: Error | null) => void;
+
+export type CallbackWithResult<T> = (
+  error?: Error | null,
+  result?: T | null
+) => void;
+
+export type KeyValuePair = [string, string | null];
+
+export type MultiCallback = (errors?: readonly (Error | null)[] | null) => void;
+
+export type MultiGetCallback = (
+  errors?: readonly (Error | null)[] | null,
+  result?: readonly KeyValuePair[]
+) => void;
+
+export type MultiRequest = {
+  keys: readonly string[];
+  callback?: MultiGetCallback;
+  keyIndex: number;
+  resolve?: (result?: readonly KeyValuePair[]) => void;
+  reject?: (error?: any) => void;
+};
+
+export type AsyncStorageHook = {
+  getItem: (callback?: CallbackWithResult<string>) => Promise<string | null>;
+  setItem: (value: string, callback?: Callback) => Promise<void>;
+  mergeItem: (value: string, callback?: Callback) => Promise<void>;
+  removeItem: (callback?: Callback) => Promise<void>;
+};
```

Comment:
It would be helpful to explicitly set the `this` type to `void` here. Users with an eslint rule of [`unbound-method`](https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/unbound-method.md) would see check failures when using destructuring the hook's response.

```suggestion
export type AsyncStorageHook = {
  getItem: (this: void, callback?: CallbackWithResult<string>) => Promise<string | null>;
  setItem: (this: void, value: string, callback?: Callback) => Promise<void>;
  mergeItem: (this: void, value: string, callback?: Callback) => Promise<void>;
  removeItem: (this: void, callback?: Callback) => Promise<void>;
};
```

I submitted a [separate PR that does that for the existing types](https://github.com/react-native-async-storage/async-storage/pull/736) while a PR like this is still in early draft stages.